context/rxjava2/src/main/java/brave/context/rxjava2/TraceContextObserver.java

```diff
@@ -25,7 +25,7 @@
   }
 
   @Override public final void onSubscribe(Disposable d) {
-    if (DisposableHelper.validate(upstream, d)) {
+    if (Util.validate(upstream, d)) {
       downstream.onSubscribe((upstream = d));
```

Comment:
The upstream disposable in an Observer should never just passed along as it may cause problems with fusion.

Reply:
can you point me to a line of doing this right? I'm concerned about inheriting a tree of internal classes, but it would still be useful to know what correct is. Also, if you have any insight on how to prove this will be wrong, that would help, too.

Reply:
https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/observers/BasicFuseableObserver.java#L66

Try instrumenting something like this: `Observable.range(1, 10).concatMap(Observable::just)`. If I understand brave correctly, the `TraceContextObserver` is sandwitched between `range` and `concatMap`, `concatMap` will receive the `QueueDisposable` of `range` and poll it directly, thus `TraceContextObserver` will never get its `onXXX` methods called.

Reply:
thanks for your patience.. I'm moving this stuff to a new PR where I can hopefully address it. I will read into this and try to understand https://github.com/openzipkin/brave/pull/814

Reply:
ps I think you are right.. I will address over there